File: src/lib/balances.ts
import { Expense } from "../types";

const EPSILON = 0.005;

/**
 * Net balance per person: positive means the household owes them money,
 * negative means they owe the household.
 */
export function computeBalances(people: string[], expenses: Expense[]): Record<string, number> {
  const balances: Record<string, number> = {};
  people.forEach((p) => (balances[p] = 0));

  for (const { paidBy, amount, split, forPerson } of expenses) {
    if (!(paidBy in balances)) continue;
    if (split === "even") {
      const perPerson = amount / people.length;
      for (const p of people) {
        if (p !== paidBy) balances[p] -= perPerson;
        else balances[p] += amount - perPerson;
      }
    } else if (forPerson && forPerson !== paidBy && forPerson in balances) {
      balances[forPerson] -= amount;
      balances[paidBy] += amount;
    }
  }

  return balances;
}

export interface Settlement {
  from: string;
  to: string;
  amount: number;
}

/** Reduce balances to a minimal set of transfers. */
export function computeSettlements(balances: Record<string, number>): Settlement[] {
  const creditors: { person: string; amount: number }[] = [];
  const debtors: { person: string; amount: number }[] = [];

  for (const [person, bal] of Object.entries(balances)) {
    if (bal > EPSILON) creditors.push({ person, amount: bal });
    else if (bal < -EPSILON) debtors.push({ person, amount: -bal });
  }

  creditors.sort((a, b) => b.amount - a.amount);
  debtors.sort((a, b) => b.amount - a.amount);

  const settlements: Settlement[] = [];
  let i = 0;
  let j = 0;
  while (i < creditors.length && j < debtors.length) {
    const c = creditors[i];
    const d = debtors[j];
    const amount = Math.min(c.amount, d.amount);
    settlements.push({ from: d.person, to: c.person, amount });
    c.amount -= amount;
    d.amount -= amount;
    if (c.amount < EPSILON) i++;
    if (d.amount < EPSILON) j++;
  }
  return settlements;
}

File: native/Roomies/Sources/Models/Balances.swift
import Foundation

private let EPSILON = 0.005

/// Net balance per person: positive means the household owes them money,
/// negative means they owe the household.
func computeBalances(people: [String], expenses: [Expense]) -> [String: Double] {
    var balances: [String: Double] = [:]
    for p in people { balances[p] = 0 }

    for expense in expenses {
        guard balances[expense.paidBy] != nil else { continue }
        switch expense.split {
        case .even:
            let perPerson = expense.amount / Double(people.count)
            for p in people {
                if p != expense.paidBy {
                    balances[p, default: 0] -= perPerson
                } else {
                    balances[p, default: 0] += expense.amount - perPerson
                }
            }
        case .full:
            if let forPerson = expense.forPerson, forPerson != expense.paidBy, balances[forPerson] != nil {
                balances[forPerson, default: 0] -= expense.amount
                balances[expense.paidBy, default: 0] += expense.amount
            }
        }
    }

    return balances
}

struct Settlement: Identifiable, Equatable {
    var from: String
    var to: String
    var amount: Double
    var id: String { "\(from)->\(to)" }
}

/// Reduce balances to a minimal set of transfers.
func computeSettlements(balances: [String: Double]) -> [Settlement] {
    var creditors: [(person: String, amount: Double)] = []
    var debtors: [(person: String, amount: Double)] = []

    for (person, bal) in balances {
        if bal > EPSILON { creditors.append((person, bal)) }
        else if bal < -EPSILON { debtors.append((person, -bal)) }
    }

    creditors.sort { $0.amount > $1.amount }
    debtors.sort { $0.amount > $1.amount }

    var settlements: [Settlement] = []
    var i = 0
    var j = 0
    while i < creditors.count && j < debtors.count {
        let amount = min(creditors[i].amount, debtors[j].amount)
        settlements.append(Settlement(from: debtors[j].person, to: creditors[i].person, amount: amount))
        creditors[i].amount -= amount
        debtors[j].amount -= amount
        if creditors[i].amount < EPSILON { i += 1 }
        if debtors[j].amount < EPSILON { j += 1 }
    }
    return settlements
}
